utils/set/sampleable_set.go

```diff
@@ -0,0 +1,231 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package set
+
+import (
+	"bytes"
+
+	stdjson "encoding/json"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/ava-labs/avalanchego/utils"
+	"github.com/ava-labs/avalanchego/utils/json"
+	"github.com/ava-labs/avalanchego/utils/math"
+	"github.com/ava-labs/avalanchego/utils/sampler"
+	"github.com/ava-labs/avalanchego/utils/wrappers"
+)
+
+var _ stdjson.Marshaler = (*Set[int])(nil)
+
+// SampleableSet is a set of elements that supports sampling.
+type SampleableSet[T comparable] struct {
+	// indices maps the element in the set to the index that it appears in
+	// elements.
+	indices  map[T]int
+	elements []T
+}
+
+// Return a new sampleable set with initial capacity [size].
+// More or less than [size] elements can be added to this set.
+// Using NewSampleableSet() rather than SampleableSet[T]{} is just an
+// optimization that can be used if you know how many elements will be put in
+// this set.
+func NewSampleableSet[T comparable](size int) SampleableSet[T] {
+	if size < 0 {
+		return SampleableSet[T]{}
+	}
+	return SampleableSet[T]{
+		indices:  make(map[T]int, size),
+		elements: make([]T, 0, size),
+	}
+}
+
+// Add all the elements to this set.
+// If the element is already in the set, nothing happens.
+func (s *SampleableSet[T]) Add(elements ...T) {
+	s.resize(2 * len(elements))
+	for _, e := range elements {
+		s.add(e)
+	}
+}
+
+// Union adds all the elements from the provided set to this set.
+func (s *SampleableSet[T]) Union(set SampleableSet[T]) {
+	s.resize(2 * set.Len())
+	for _, e := range set.elements {
+		s.add(e)
+	}
+}
+
+// Difference removes all the elements in [set] from [s].
+func (s *SampleableSet[T]) Difference(set SampleableSet[T]) {
+	for _, e := range set.elements {
+		s.remove(e)
+	}
+}
+
+// Contains returns true iff the set contains this element.
+func (s SampleableSet[T]) Contains(e T) bool {
```

Comment:
why are some fns defined on the struct receiver and others are defined on the pointer receiver?